File: agent/robotscope_agent/protocol.py
"""RobotScope live WebSocket protocol v0.1."""

from __future__ import annotations

import base64
import json
from typing import Any

LIVE_PROTOCOL_VERSION = "robotscope.live.v0.1"
AGENT_NAME = "robotscope-ros2-agent"


def encode_payload(data: bytes) -> str:
    return base64.b64encode(data).decode("ascii")


def session_message(
    start_ns: int,
    topics: list[dict[str, str]],
    *,
    publish_topics: list[str] | None = None,
    service_call_services: list[str] | None = None,
) -> str:
    payload: dict[str, Any] = {
        "type": "session",
        "protocol": LIVE_PROTOCOL_VERSION,
        "agent": AGENT_NAME,
        "start_ns": start_ns,
        "topics": topics,
    }
    capabilities: dict[str, Any] = {}
    if publish_topics:
        capabilities["command_publish"] = publish_topics
    if service_call_services:
        capabilities["command_service_call"] = service_call_services
    if capabilities:
        payload["capabilities"] = capabilities
    return json.dumps(payload)


def channel_message(channel: dict[str, Any]) -> str:
    return json.dumps({"type": "channel", "channel": channel})


def data_message(
    channel_id: int,
    log_time_ns: int,
    data: bytes,
    publish_time_ns: int | None = None,
    sequence: int | None = None,
) -> str:
    payload: dict[str, Any] = {
        "type": "message",
        "channel_id": channel_id,
        "log_time_ns": log_time_ns,
        "data_b64": encode_payload(data),
    }
    if publish_time_ns is not None:
        payload["publish_time_ns"] = publish_time_ns
    if sequence is not None:
        payload["sequence"] = sequence
    return json.dumps(payload)


def status_message(
    phase: str,
    message: str | None = None,
    *,
    topics_subscribed: int | None = None,
    topics_pending: int | None = None,
) -> str:
    payload: dict[str, Any] = {"type": "status", "phase": phase}
    if message:
        payload["message"] = message
    if topics_subscribed is not None:
        payload["topics_subscribed"] = topics_subscribed
    if topics_pending is not None:
        payload["topics_pending"] = topics_pending
    return json.dumps(payload)


def error_message(message: str) -> str:
    return json.dumps({"type": "error", "message": message})


def command_publish_result_message(
    ok: bool,
    message: str,
    *,
    topic: str | None = None,
) -> str:
    payload: dict[str, Any] = {
        "type": "command.publish_result",
        "ok": ok,
        "message": message,
    }
    if topic:
        payload["topic"] = topic
    return json.dumps(payload)


def command_service_result_message(
    ok: bool,
    message: str,
    *,
    service: str | None = None,
    success: bool | None = None,
) -> str:
    payload: dict[str, Any] = {
        "type": "command.service_result",
        "ok": ok,
        "message": message,
    }
    if service:
        payload["service"] = service
    if success is not None:
        payload["success"] = success
    return json.dumps(payload)

File: packages/robotscope-core/src/live/protocol.ts
/** RobotScope live agent WebSocket protocol v0.1 */

import type { FibonacciActionGoal } from "./action-gateway.js";
import type { TwistVelocityCommand } from "./command-gateway.js";

export const LIVE_PROTOCOL_VERSION = "robotscope.live.v0.1" as const;

export interface LiveChannelDefinition {
  id: number;
  topic: string;
  schema: string;
  encoding: "ros2msg" | "ros2idl" | string;
  /** ROS message definition text (ros2msg). */
  definition: string;
}

export interface LiveSessionMessage {
  type: "session";
  protocol: typeof LIVE_PROTOCOL_VERSION;
  agent: string;
  start_ns: number;
  topics: Array<{ name: string; schema: string }>;
  capabilities?: {
    command_publish?: string[];
    command_service_call?: string[];
    command_service_types?: Record<string, string>;
    command_action_send_goal?: string[];
  };
}

export interface LiveChannelMessage {
  type: "channel";
  channel: LiveChannelDefinition;
}

export interface LiveDataMessage {
  type: "message";
  channel_id: number;
  log_time_ns: number;
  publish_time_ns?: number;
  sequence?: number;
  /** Base64-encoded CDR payload. */
  data_b64: string;
}

export interface LiveStatusMessage {
  type: "status";
  phase: "connecting" | "waiting_for_topics" | "ready" | "streaming" | "error";
  message?: string;
  topics_subscribed?: number;
  topics_pending?: number;
}

export interface LiveErrorMessage {
  type: "error";
  message: string;
}

export interface LiveCommandPublishResultMessage {
  type: "command.publish_result";
  ok: boolean;
  topic?: string;
  message: string;
}

export interface LiveCommandServiceResultMessage {
  type: "command.service_result";
  ok: boolean;
  service?: string;
  message: string;
  success?: boolean;
}

export interface LiveCommandActionResultMessage {
  type: "command.action_result";
  ok: boolean;
  action?: string;
  message: string;
  goal_accepted?: boolean;
}

export interface LiveCommandActionFeedbackMessage {
  type: "command.action_feedback";
  action: string;
  sequence: number[];
}

export interface LiveCommandActionOutcomeMessage {
  type: "command.action_outcome";
  action: string;
  ok: boolean;
  status: "succeeded" | "aborted" | "canceled" | "failed";
  sequence: number[];
  message?: string;
}

export interface LiveCommandActionCancelResultMessage {
  type: "command.action_cancel_result";
  ok: boolean;
  action?: string;
  message: string;
  cancel_accepted?: boolean;
}

export type LiveServerMessage =
  | LiveSessionMessage
  | LiveChannelMessage
  | LiveDataMessage
  | LiveStatusMessage
  | LiveErrorMessage
  | LiveCommandPublishResultMessage
  | LiveCommandServiceResultMessage
  | LiveCommandActionResultMessage
  | LiveCommandActionFeedbackMessage
  | LiveCommandActionOutcomeMessage
  | LiveCommandActionCancelResultMessage;

export interface LiveCommandPublishClientMessage {
  type: "command.publish";
  topic: string;
  schema: string;
  twist?: TwistVelocityCommand;
  zero_twist?: boolean;
  data_b64?: string;
}

export interface LivePingMessage {
  type: "ping";
}

export interface LiveCommandServiceCallClientMessage {
  type: "command.service_call";
  service: string;
  schema: string;
  trigger?: boolean;
  data?: boolean;
}

export interface LiveCommandActionSendGoalClientMessage {
  type: "command.action_send_goal";
  action: string;
  schema: string;
  fibonacci?: FibonacciActionGoal;
  preempt?: boolean;
}

export interface LiveCommandActionCancelGoalClientMessage {
  type: "command.action_cancel_goal";
  action: string;
}

export type LiveClientMessage =
  | LivePingMessage
  | LiveCommandPublishClientMessage
  | LiveCommandServiceCallClientMessage
  | LiveCommandActionSendGoalClientMessage
  | LiveCommandActionCancelGoalClientMessage;

export function parseLiveServerMessage(raw: string): LiveServerMessage | null {
  try {
    const value = JSON.parse(raw) as LiveServerMessage;
    if (!value || typeof value !== "object" || !("type" in value)) {
      return null;
    }
    return value;
  } catch {
    return null;
  }
}

export function encodeLiveClientMessage(message: LiveClientMessage): string {
  return JSON.stringify(message);
}

export function decodeBase64Payload(data_b64: string): Uint8Array {
  if (typeof globalThis.atob === "function") {
    const binary = globalThis.atob(data_b64);
    const bytes = new Uint8Array(binary.length);
    for (let i = 0; i < binary.length; i += 1) {
      bytes[i] = binary.charCodeAt(i);
    }
    return bytes;
  }

  // Node fallback for tests/scripts importing core helpers.
  const buffer = Buffer.from(data_b64, "base64");
  return new Uint8Array(buffer.buffer, buffer.byteOffset, buffer.byteLength);
}
